win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+
+#undef ENABLE_MOCKS
+
+#include "azure_c_pal/file.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+TEST_DEFINE_ENUM_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_RESULT)
+IMPLEMENT_UMOCK_C_ENUM_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES)
+
+TEST_DEFINE_ENUM_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_RESULT)
+IMPLEMENT_UMOCK_C_ENUM_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+static FILE_HANDLE get_file_handle(const char* filename, PTP_WIN32_IO_CALLBACK* captured_callback)
+{
+    ///arrange
+    EXECUTION_ENGINE_HANDLE execution_engine = (EXECUTION_ENGINE_HANDLE)41;
+    HANDLE handle = (HANDLE)42;
+    PTP_POOL ptp_pool = (PTP_POOL)43;
+    PTP_CALLBACK_ENVIRON captured_pcbe;
+    PTP_CLEANUP_GROUP ptp_cleanup_group = (PTP_CLEANUP_GROUP)44;
+    PTP_IO ptp_io = (PTP_IO)45;
+
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateFileA(filename, GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ, NULL, OPEN_ALWAYS, FILE_FLAG_OVERLAPPED | FILE_FLAG_WRITE_THROUGH, NULL))
+        .SetReturn(handle);
+    STRICT_EXPECTED_CALL(mock_SetFileCompletionNotificationModes(handle, FILE_SKIP_COMPLETION_PORT_ON_SUCCESS))
+        .SetReturn(TRUE);
+    STRICT_EXPECTED_CALL(mock_InitializeThreadpoolEnvironment(IGNORED_ARG))
+        .CaptureArgumentValue_pcbe(&captured_pcbe);
+    STRICT_EXPECTED_CALL(execution_engine_win32_get_threadpool(execution_engine))
+        .SetReturn(ptp_pool);
+    STRICT_EXPECTED_CALL(mock_SetThreadpoolCallbackPool(IGNORED_ARG, ptp_pool))
+        .ValidateArgumentValue_pcbe(&captured_pcbe);
+    STRICT_EXPECTED_CALL(mock_CreateThreadpoolCleanupGroup())
+        .SetReturn(ptp_cleanup_group);
+    STRICT_EXPECTED_CALL(mock_SetThreadpoolCallbackCleanupGroup(IGNORED_ARG, ptp_cleanup_group, IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateThreadpoolIo(handle, IGNORED_ARG, NULL, IGNORED_ARG))
+        .CaptureArgumentValue_pcbe(&captured_pcbe)
+        .CaptureArgumentValue_pfnio(captured_callback)
+        .SetReturn(ptp_io);
+
+    ///act
+    FILE_HANDLE file_handle = file_create(execution_engine, filename, NULL, NULL);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(file_handle);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    umock_c_reset_all_calls();
+    return file_handle;
+}
+
+static void get_callback(FILE_CB user_callback, void* user_context, uint32_t size, PTP_WIN32_IO_CALLBACK* captured_callback, LPOVERLAPPED* captured_ov)
+{
+    ///arrange
+    FILE_HANDLE file_handle = get_file_handle("test_file.txt", captured_callback);
+    unsigned char buf[10];
+    HANDLE h_event = (HANDLE)44;
+    uint64_t position = 5;

```

Comment:
>position [](start = 13, length = 8)

I'd think you would want to check the offset on the captured overlapped, which means the caller should know what this is, which means I think again it should be a param here.